library/psa_crypto.c

```diff
@@ -713,118 +713,187 @@ static psa_status_t psa_import_rsa_key( psa_key_slot_t *slot,
+        if( status != PSA_SUCCESS )
+            goto exit;
+    }
+    else
+    {
+        /* Load the secret value. */
+        status = mbedtls_to_psa_error(
+            mbedtls_ecp_read_key( ecp->grp.id,
+                                  ecp,
+                                  slot->data.key.data,
+                                  slot->data.key.bytes ) );
 
-    /* Load the public value. */
-    status = mbedtls_to_psa_error(
-        mbedtls_ecp_point_read_binary( &ecp->grp, &ecp->Q,
-                                       data, data_length ) );
+        if( status != PSA_SUCCESS )
+            goto exit;
+        /* Validate the private key. */
+        status = mbedtls_to_psa_error(
+            mbedtls_ecp_check_privkey( &ecp->grp, &ecp->d ) );
+        if( status != PSA_SUCCESS )
+            goto exit;
+    }
+exit:
     if( status != PSA_SUCCESS )
-        goto exit;
+        mbedtls_ecp_keypair_free( ecp );
+    return status;
+}
 
-    /* Check that the point is on the curve. */
-    status = mbedtls_to_psa_error(
-        mbedtls_ecp_check_pubkey( &ecp->grp, &ecp->Q ) );
-    if( status != PSA_SUCCESS )
-        goto exit;
+static psa_status_t psa_export_ecp_key( psa_key_type_t type,
+                                        mbedtls_ecp_keypair *ecp,
+                                        uint8_t *data,
+                                        size_t data_size,
+                                        size_t *data_length )
+{
+    psa_status_t status;
 
-    *p_ecp = ecp;
-    return( PSA_SUCCESS );
+    if( PSA_KEY_TYPE_IS_PUBLIC_KEY( type ) )
+    {
+        /* Check whether the public part is loaded */
+        if( mbedtls_ecp_is_zero( &ecp->Q ) )
+        {
+            /* Calculate the public key */
+            status = mbedtls_to_psa_error(
+                mbedtls_ecp_mul( &ecp->grp, &ecp->Q, &ecp->d, &ecp->grp.G,
+                                 mbedtls_ctr_drbg_random, &global_data.ctr_drbg ) );
+            if( status != PSA_SUCCESS )
+                return status;
+        }
 
-exit:
-    if( ecp != NULL )
+        return( mbedtls_to_psa_error(
+                    mbedtls_ecp_point_write_binary( &ecp->grp, &ecp->Q,
+                                                    MBEDTLS_ECP_PF_UNCOMPRESSED,
+                                                    data_length,
+                                                    data,
+                                                    data_size ) ) );
+    }
+    else
     {
-        mbedtls_ecp_keypair_free( ecp );
-        mbedtls_free( ecp );
+        if( data_size < PSA_BITS_TO_BYTES(ecp->grp.nbits) )
+            return( PSA_ERROR_BUFFER_TOO_SMALL );
+
+        status = mbedtls_to_psa_error(
+                    mbedtls_ecp_write_key( ecp,
+                                           data,
+                                           PSA_BITS_TO_BYTES(ecp->grp.nbits) ) );
+        if( status == PSA_SUCCESS )
+        {
```

Comment:
I don't see where the API contract promises to zero-out the buffer beyond what it has indicated it has written to?
If it's a requirement that all these APIs write the full buffer it gets passed, then there's a ton of other places that should be updated to `memset` the unused buffer space.

If I had to take a guess, the fact that the `memset` got introduced at all is an artifact of using the pk_write functions further down in the previous content of `internal_export_key`. There, a memset is needed, since the (potentially private) key material got first loaded to the end of the buffer, and then moved over, meaning you need to overwrite the end of the buffer to avoid leaking key material. This is not a concern when not using PK calls, since there never gets written anything outside of the boundary indicated by `*data_length`.

I don't think this requires its own commit, since the previous memset was never a part of the API contract, and, as I pointed out here, probably misguided in the first place.

Reply:
API functions must not leave confidential data lying around in unused memory: confidential data that isn't an output to the caller must be wiped. In Mbed TLS, we use `mbedtls_platform_zeroize` for that. This protects against data leaks such as Heartbleed.

In addition, even non-confidential data can cause trouble if the application doesn't work perfectly. For example, if the application doesn't check the return status properly, it may end up using the content of a buffer with indeterminate content. So we try to set unused parts of output buffers (which is the whole buffer if the API function fails) to “safe values”, [as recommended in the PSA specification](https://armmbed.github.io/mbed-crypto/html/overview/implementation.html#safe-outputs-on-error). This typically means zeroing the unused part. For this use case, `memset` is enough, we don't need `zeroize`, since it's ok if the compiler optimizes it away (because it'll only do that if it determines that the caller doesn't use the content).

Many API functions zero out the whole output buffer near their entry point. It's easy and is a small cost if the output buffer is small. For the pk export, it's more complicated because as you note `pk_write` writes from the end and we needed to move the key around.

I haven't yet reviewed the whole code, but I think `psa_internal_export_key` is now missing a `memset` to wipe the unused part of the output buffer.

Reply:
For `psa_internal_export_key`:
* The RSA key export function uses the PK module, and comes with a memset due to having to move the key around
* The ECP key export uses the mbedTLS MPI write functions, which either succeed or don't write anything, meaning that the buffer we return to the application only contains the bytes that the application passed us in the first place.
* The raw key output comes with a memset, though I still think it's unnecessary given that the buffer contents that are being memset'ed have been passed to us by the application, meaning that the 'potentially leaky' contents of the buffer are the exact same bytes as were passed to us in the first place.